lib/pbench/server/api/resources/query_apis/search_api.py

```diff
@@ -0,0 +1,221 @@
+from http import HTTPStatus
+from flask import jsonify
+from logging import Logger
+from typing import Any, AnyStr, Dict
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources.query_apis import (
+    CONTEXT,
+    ElasticBase,
+    Schema,
+    Parameter,
+    ParamType,
+    PostprocessError,
+)
+
+
+class SearchAPI(ElasticBase):
+    """
+    Create a search request based on given query parameter.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("user", ParamType.USER, required=False),
+                Parameter("start", ParamType.DATE, required=True),
+                Parameter("end", ParamType.DATE, required=True),
+                Parameter("search_term", ParamType.STRING, required=False),
+                Parameter("fields", ParamType.LIST, required=False),
+            ),
+        )
+
+    def flatten_json(self, es_json: Dict[AnyStr, Any]) -> Dict[AnyStr, Any]:
+        """
+        Utility function for flattening the json.
+        We convert a nested data layer object into a new object with only one layer of key/value pairs.
+        Objects such as parent -> child are flattened as parent.child and parent -> array -> [obj1,objN]
+        are flattened as parent.array.0 = obj1 parent.array.N = objN
+        json_in =
+            {
+                    "@timestamp": "2021-07-12T14:08:46.563303",
+                    "run": {
+                        "controller": "dhcp31-171.perf.lab.eng.bos.redhat.com",
+                        "name": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46"
+                    },
+                    "host_tools_info": [
+                        {
+                            "hostname": [{"test1":"dhcp31-171"}, {"test2":"dhcp31-195"}]
+                        }
+                    ]
+                }
+        json_out =
+            {
+                '@timestamp': '2021-07-12T14:08:46.563303',
+                'run.controller': 'dhcp31-171.perf.lab.eng.bos.redhat.com',
+                'run.name': 'pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46',
+                'host_tools_info.hostname.test1': 'dhcp31-171',
+                'host_tools_info.hostname.test2': 'dhcp31-195'
+            }
+        """
+        if type(es_json) is not dict:
+            raise PostprocessError(
+                HTTPStatus.INTERNAL_SERVER_ERROR,
+                "Elasticsearch _source value is not a json",
+            )
+
+        def x(es_json: Any, prefix="", out={}):
+            if type(es_json) is dict:
+                for key in es_json:
+                    while prefix[-2:-1] == ".":
+                        prefix = prefix[:-1]
```

Comment:
> For instance, if the input were
> 
> ```
> {
>   "inlist": [0, 1, 2, 3]
> }
> ```
> 
> Then I think we would want the output to look exactly like the input, and _not_ like
> 
> ```
> {
>   "inlist.0": 0,
>   "inlist.1": 1,
>   "inlist.2": 2,
>   "inlist.3": 3
> }
> ```

For the example that you shown here is actually how the _flattened_ view of that dict looks like. Flattening by definition means to convert a nested data layer object into a new object with only _one layer of key/value_ pairs. If we were to return `{ "inlist": [0, 1, 2, 3] }` as it is then this algorithm won't work at the moment. 

Also this
```
{
  "inlist": [ { "inlist0": [0, 1]}, { "inlist0": [0, 1]}, ]
}
```
should convert into 
```
{
  "inlist.inlist0": [ [0, 1], [0, 1] ]
}
```
isn't it? because inner 2 lists were separate values.
But going back to your first example, my concern would be how do we handle a case where the elements of list are of different data types. For example, 
```
{
 "inlist": [0, 1, {"key": "value"}, 3]
}
```
do we want this to be like this:
```
{
 "inlist": [0, 1, 3],
 "inlist.key": "value"
}
```
because this seems wrong isn't it?

Reply:
> should convert into
{
  "inlist.inlist0": [ [0, 1], [0, 1] ]
}
isn't it? because inner 2 lists were separate values.

Based on the code that you've written, I've been assuming that lists are not "flat" (I'm proposing that we change to ignore that detail at the leaves...), so a list containing things that aren't scalar would need to be flattened.

>{
 "inlist": [0, 1, 3],
 "inlist.key": "value"
}
because this seems wrong isn't it?

Whether that seems wrong or not depends on your perspective or requirements.  Aside from the fact that heterogenous lists are weird (i.e., they are tuple-wannabes), a list containing an object is not "flat".  The fact that in your example here the requestor can access the scalars via `"inlist"` and the sub-object values via `"inlist.key"` relieves the feeling of wrongness for me.  (One can tell by looking that `"inlist.key"` is related and subordinate to `"inlist"`, thus that relationship is preserved; however, while one cannot reach the value of `"inlist.key"` by iterating over the values of `"inlist"`, that bothers me less than it might -- if one really wants to inspect the data, we shouldn't be flattening it -- flattening is supposed to (I think...) ease access to specific values, not present the data in a particular structure.)

I think the question that you're starting to ask here is, should we be flattening lists at all, or should we be leaving them intact and flattening only cleanly-nested _objects_?

Reply:
We really have to address two separate issues, here:

1. How does the dashboard code deal with the results, and what are the actual requirements for page construction?
2. How can the Pbench server, as a client-independent architectural layer between client and backend database, best represent the results of a general search to a client? Flattening discards structural information, but to a large extent the structure we're representing is unnaturally tied to the current Elasticsearch representation and, in principle, we'd like to isolate the client from current implementation details.

Unfortunately it's difficult to separate these issues in a casual discussion. The most critical part, though, is to work out exactly what the dashboard needs to build the table, and then we can think about how we want to provide that; whether all from the back end as we're trying to do here, or with UI processing on top of a more general back end response.

Reply:
See also [this comment](https://github.com/distributed-system-analysis/pbench/pull/2381#discussion_r688172586).